src/graphql.js

```diff
@@ -628,4 +628,258 @@ function authAPIcall(query,variables,callback,cacheKey,timeFresh,useLocalStorage
+let apiResetLimit;
+
+/** Provides default arguments for an {@link anilistAPI()} request */
+class QueryOptions{
+	/**
+	 * Creates an arguments object
+	 * @param {object} [args={}] - Contains the options which will override the default options for the request.
+	 */
+	constructor(args={}){
+		this.variables = {};
+		this.cacheKey = null;
+		this.duration = null;
+		this.overwrite = false;
+		this.auth = false;
+		this.internal = false;
+		Object.assign(this, args);
+	}
+}
+
+/** Configures options for a network request */
+class RequestOptions{
+	constructor(query, variables, auth, internal){
+		this.method = "POST",
+		this.headers = new Headers({
+			"Content-Type": "application/json",
+			"Accept": "application/json"
+		}),
+		this.body = JSON.stringify({
+			"query": query,
+			"variables": variables
+		})
+		this.#isAuth(auth)
+		this.#isInternal(internal)
+	}
+	#isAuth(auth){
+		if(auth === true && useScripts.accessToken){
+			this.headers.set("Authorization", "Bearer " + useScripts.accessToken)
+		}
+	}
+	#isInternal(internal){
+		if(internal === true){
+			if(al_token){
+				this.headers.set("x-csrf-token", al_token)
+			}
+			this.headers.set("schema", "internal")
+		}
+	}
+}
+
+/**
+ * Updates the variables tracking the API request limit
+ * @param {Response} res - The Response object from a network request.
+ */
+function updateLimit(res){
+	if(res.headers.has("x-ratelimit-limit")){
+		APIlimit = res.headers.get("x-ratelimit-limit");
+	}
+	if(res.headers.has("x-ratelimit-remaining")){
+		APIcallsUsed = APIlimit - res.headers.get("x-ratelimit-remaining");
+	}
+}
+
+/**
+ * Checks the API cache for an existing item and returns it if it has not expired
+ * @param {string} key - The key to check for in the datastore.
+ * @returns {Promise<object>} The cached data if found.
+ */
+async function checkCache(key){
+	const item = await apiCache.getItem(key);
+	if(item){
+		if(!item.expiresAt || (NOW() < item.expiresAt)){
+			return item.data;
+		}
+		else{
+			return apiCache.removeItem(key);
+		}
+	}
+	return;
+}
+
+/**
+ * Iterates the API cache for expired items and removes them
+ * @returns {Promise}
+ */
+async function flushCache(){
+	return apiCache.iterate((item, key) => {
+		if(NOW() > item.expiresAt){
+			apiCache.removeItem(key);
+		}
+	})
+}
+
+/**
+ * Saves data to the API cache
+ * @param {string} key - The key used to store the data.
+ * @param {object} data - The data to store.
+ * @param {number} [duration] - The length of time to store in seconds.
+ * @return {Promise}
+ */
+async function saveCache(key, data, duration){
+	const saltedHam = {
+		data: data,
+		createdAt: NOW(),
+		expiresAt: duration ? NOW() + duration*1000 : undefined
+	};
+	try{
+		return apiCache.setItem(key, saltedHam);
+	}
+	catch(e){
+		if(e.name === "QuotaExceededError"){
+			console.error("Persistent storage quota exceeded. Attempting to purge expired items.")
+			try{
+				return flushCache();
```

Comment:
But what if purging expired items is not enough?

One could imagine a more aggressive pruning strategy, based on item size, time, and number of previous cache hits that attempts to remove even more items if the first round of pruning is not enough.

Reply:
I mean, at that point I wonder if the user would have enough free disk space for the pruning to even make a difference.

IndexedDB is generally supposed to have a much larger limit than the 5MB that Localstorage has.
https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Browser_storage_limits_and_eviction_criteria#storage_limits

I was previously considering dropping the store entirely as a last resort with [`apiCache.clear()`](https://localforage.github.io/localForage/#data-api-clear)

Reply:
That sounds better.

Reply:
added in f97a77c